File: packages/sandboxd/src/process/lifecycle.rs
//! Runtime-client process start/stop lifecycle helpers.
//!
//! This module owns environment merging, child spawn, readiness waiting, and
//! shutdown signal policy for generic runtime processes.

use std::collections::BTreeMap;
use std::process::{Child, Command, Stdio};

use nix::errno::Errno;
use nix::sys::signal::{Signal, kill as send_signal};
use nix::unistd::Pid;

use crate::cgroups::{attach_pid_to_scope, kill_scope};
use crate::process::*;
use crate::runtime::{RuntimeClientProcessReadiness, RuntimeClientProcessStopSignal};
use crate::time::{Clock, Sleeper};

/// Merges client-wide environment variables with any process-local overrides.
pub(super) fn merge_runtime_client_process_env(
    runtime_env: &BTreeMap<String, String>,
    runtime_client_env: &BTreeMap<String, String>,
    process_command_env: Option<&BTreeMap<String, String>>,
) -> Option<BTreeMap<String, String>> {
    if runtime_env.is_empty()
        && runtime_client_env.is_empty()
        && process_command_env.is_none_or(BTreeMap::is_empty)
    {
        return None;
    }

    let mut merged_env = runtime_env.clone();
    for (key, value) in runtime_client_env {
        merged_env.insert(key.clone(), value.clone());
    }
    if let Some(process_command_env) = process_command_env {
        for (key, value) in process_command_env {
            merged_env.insert(key.clone(), value.clone());
        }
    }

    Some(merged_env)
}

/// Spawns one runtime client child process with stdio detached from `sandboxd`.
pub(super) fn start_runtime_client_process(
    process_spec: &RuntimeClientProcessSpec,
) -> Result<RunningRuntimeClientProcess, String> {
    let (child, output_capture) = spawn_runtime_client_child(process_spec)?;

    Ok(RunningRuntimeClientProcess {
        spec: process_spec.clone(),
        child: Arc::new(Mutex::new(child)),
        output_capture,
        platform_scope: None,
    })
}

pub(super) fn start_scoped_runtime_client_process(
    process_spec: &RuntimeClientProcessSpec,
    platform_scope: RuntimeClientProcessPlatformScope,
) -> Result<RunningRuntimeClientProcess, String> {
    let (child, output_capture) = match spawn_runtime_client_child(process_spec) {
        Ok(started_process) => started_process,
        Err(error) => {
            if let Err(cleanup_error) =
                kill_runtime_client_process_platform_scope_by_scope(&platform_scope)
            {
                return Err(format!(
                    "{error}; platform scope cleanup failed: {cleanup_error}"
                ));
            }
            return Err(error);
        }
    };
    let process_id = child.id();
    if let Err(error) = attach_pid_to_scope(&platform_scope.scope_paths, process_id) {
        let mut failed_process = RunningRuntimeClientProcess {
            spec: process_spec.clone(),
            child: Arc::new(Mutex::new(child)),
            output_capture,
            platform_scope: Some(platform_scope.clone()),
        };
        let _ = signal_runtime_client_process(
            &mut failed_process,
            RuntimeClientProcessStopSignal::Sigkill,
        );
        if let Err(cleanup_error) =
            kill_runtime_client_process_platform_scope_by_scope(&platform_scope)
        {
            return Err(format!(
                "{error}; platform scope cleanup failed: {cleanup_error}"
            ));
        }
        return Err(error.to_string());
    }

    platform_scope.registry.replace_scope(
        platform_scope.registry_key.clone(),
        platform_scope.process_key.clone(),
        platform_scope.scope_paths.clone(),
        process_id,
    )?;

    Ok(RunningRuntimeClientProcess {
        spec: process_spec.clone(),
        child: Arc::new(Mutex::new(child)),
        output_capture,
        platform_scope: Some(platform_scope),
    })
}

pub(super) fn spawn_runtime_client_child(
    process_spec: &RuntimeClientProcessSpec,
) -> Result<(Child, ProcessOutputCapture), String> {
    let command = process_spec
        .command
        .args
        .first()
        .ok_or_else(|| "process command args must not be empty".to_string())?;
    let mut child_command = Command::new(command);
    child_command
        .args(&process_spec.command.args[1..])
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    if let Some(cwd) = process_spec.command.cwd.as_deref() {
        child_command.current_dir(cwd);
    }
    if let Some(env) = &process_spec.command.env {
        child_command.envs(env);
    }

    let mut child = child_command
        .spawn()
        .map_err(|error| format!("failed to start process command: {error}"))?;

    let output_capture = ProcessOutputCapture::new();
    if let Some(stdout) = child.stdout.take() {
        output_capture.register_capture_thread(spawn_output_capture_thread(
            stdout,
            output_capture.clone(),
            OutputStream::Stdout,
        ));
    }
    if let Some(stderr) = child.stderr.take() {
        output_capture.register_capture_thread(spawn_output_capture_thread(
            stderr,
            output_capture.clone(),
            OutputStream::Stderr,
        ));
    }

    Ok((child, output_capture))
}

/// Waits until one process either becomes ready, exits early, or times out.
pub(super) fn wait_for_runtime_client_process_readiness(
    process: &mut RunningRuntimeClientProcess,
    clock: &dyn Clock,
    sleeper: &dyn Sleeper,
) -> Result<(), String> {
    let mut child = process
        .child
        .lock()
        .expect("runtime client child lock should not be poisoned");
    wait_for_runtime_client_process_readiness_with(&process.spec, &mut child, clock, sleeper)
}

pub(super) fn wait_for_runtime_client_process_readiness_with(
    process_spec: &RuntimeClientProcessSpec,
    child: &mut Child,
    clock: &dyn Clock,
    sleeper: &dyn Sleeper,
) -> Result<(), String> {
    let timeout_ms = match &process_spec.readiness {
        RuntimeClientProcessReadiness::None => return Ok(()),
        RuntimeClientProcessReadiness::Tcp { timeout_ms, .. }
        | RuntimeClientProcessReadiness::Http { timeout_ms, .. }
        | RuntimeClientProcessReadiness::Ws { timeout_ms, .. } => *timeout_ms,
    };
    let deadline_ms = clock.now_ms().saturating_add(timeout_ms);

    loop {
        if let Some(status) = child
            .try_wait()
            .map_err(|error| format!("failed to poll process exit: {error}"))?
        {
            return Err(describe_process_exit(status));
        }

        match check_runtime_client_process_readiness_from_spec(process_spec) {
            Ok(()) => return Ok(()),
            Err(error) if clock.now_ms() >= deadline_ms => {
                return Err(format!(
                    "timed out after {timeout_ms}ms waiting for readiness: {error}"
                ));
            }
            Err(_) => {}
        }

        sleeper.sleep(DEFAULT_PROCESS_READINESS_POLL_INTERVAL);
    }
}

pub(super) fn check_runtime_client_process_readiness_from_spec(
    process_spec: &RuntimeClientProcessSpec,
) -> Result<(), String> {
    match &process_spec.readiness {
        RuntimeClientProcessReadiness::None => Ok(()),
        RuntimeClientProcessReadiness::Tcp { host, port, .. } => check_tcp_readiness(host, *port),
        RuntimeClientProcessReadiness::Http {
            url,
            expected_status,
            ..
        } => check_http_readiness(url, *expected_status),
        RuntimeClientProcessReadiness::Ws { url, .. } => check_ws_readiness(url),
    }
}

pub(super) fn check_codex_app_server_post_start_readiness(
    process_spec: &RuntimeClientProcessSpec,
) -> Result<(), String> {
    if let RuntimeClientProcessReadiness::Ws { url, .. } = &process_spec.readiness {
        let ready_url = codex_app_server_readyz_url(url)?;
        return check_http_readiness_with_timeout(
            &ready_url,
            200,
            CODEX_APP_SERVER_POST_START_READINESS_TIMEOUT,
        );
    }

    check_runtime_client_process_readiness_from_spec(process_spec)
}

pub(super) fn readiness_type(process_spec: &RuntimeClientProcessSpec) -> &'static str {
    match &process_spec.readiness {
        RuntimeClientProcessReadiness::None => "none",
        RuntimeClientProcessReadiness::Tcp { .. } => "tcp",
        RuntimeClientProcessReadiness::Http { .. } => "http",
        RuntimeClientProcessReadiness::Ws { .. } => "ws",
    }
}

pub(super) fn readiness_target(process_spec: &RuntimeClientProcessSpec) -> String {
    match &process_spec.readiness {
        RuntimeClientProcessReadiness::None => "".to_string(),
        RuntimeClientProcessReadiness::Tcp { host, port, .. } => format!("{host}:{port}"),
        RuntimeClientProcessReadiness::Http { url, .. } => url.clone(),
        RuntimeClientProcessReadiness::Ws { url, .. } => url.clone(),
    }
}

pub(super) fn readiness_timeout_ms(process_spec: &RuntimeClientProcessSpec) -> u64 {
    match &process_spec.readiness {
        RuntimeClientProcessReadiness::None => 0,
        RuntimeClientProcessReadiness::Tcp { timeout_ms, .. }
        | RuntimeClientProcessReadiness::Http { timeout_ms, .. }
        | RuntimeClientProcessReadiness::Ws { timeout_ms, .. } => *timeout_ms,
    }
}

/// Stops all previously started processes, reporting every shutdown failure together.
pub(super) fn stop_started_processes(
    processes: &mut [RunningRuntimeClientProcess],
    clock: &dyn Clock,
    sleeper: &dyn Sleeper,
) -> Result<(), String> {
    let mut stop_errors = Vec::new();

    for process in processes.iter_mut().rev() {
        if let Err(error) = stop_runtime_client_process(process, clock, sleeper) {
            stop_errors.push(format!("processKey={}: {error}", process.spec.process_key));
        }
    }

    if stop_errors.is_empty() {
        return Ok(());
    }

    Err(stop_errors.join("; "))
}

/// Applies one process stop policy, including optional graceful termination first.
pub(super) fn stop_runtime_client_process(
    process: &mut RunningRuntimeClientProcess,
    clock: &dyn Clock,
    sleeper: &dyn Sleeper,
) -> Result<(), String> {
    if process_has_exited(process)? {
        process.output_capture.finish_capture_threads();
        return Ok(());
    }

    let deadline_ms = clock.now_ms().saturating_add(process.spec.stop.timeout_ms);
    signal_runtime_client_process(process, process.spec.stop.signal)?;

    if matches!(
        process.spec.stop.signal,
        RuntimeClientProcessStopSignal::Sigterm
    ) {
        let grace_period_ms = process.spec.stop.grace_period_ms.unwrap_or(0);
        if grace_period_ms > 0 {
            if wait_for_runtime_client_process_exit(process, grace_period_ms, clock, sleeper)
                .is_ok()
            {
                process.output_capture.finish_capture_threads();
                return Ok(());
            }

            signal_runtime_client_process(process, RuntimeClientProcessStopSignal::Sigkill)?;
        }
    }

    let remaining_ms = deadline_ms.saturating_sub(clock.now_ms());
    let result = wait_for_runtime_client_process_exit(process, remaining_ms, clock, sleeper);
    if result.is_ok() {
        process.output_capture.finish_capture_threads();
    }
    result
}

pub(super) fn kill_runtime_client_process_platform_scope(
    process: &RunningRuntimeClientProcess,
) -> Result<(), String> {
    let Some(platform_scope) = &process.platform_scope else {
        return Ok(());
    };

    kill_runtime_client_process_platform_scope_by_scope(platform_scope)
}

pub(super) fn kill_runtime_client_process_platform_scope_by_scope(
    platform_scope: &RuntimeClientProcessPlatformScope,
) -> Result<(), String> {
    kill_scope(&platform_scope.scope_paths).map_err(|error| error.to_string())?;
    platform_scope
        .registry
        .remove_scope(&platform_scope.registry_key)?;
    Ok(())
}

/// Checks whether a child process has already exited without blocking.
pub(super) fn process_has_exited(
    process: &mut RunningRuntimeClientProcess,
) -> Result<bool, String> {
    process
        .child
        .lock()
        .expect("runtime client child lock should not be poisoned")
        .try_wait()
        .map(|status| status.is_some())
        .map_err(|error| format!("failed to poll process exit: {error}"))
}

/// Waits for one child process to exit before the given timeout elapses.
pub(super) fn wait_for_runtime_client_process_exit(
    process: &mut RunningRuntimeClientProcess,
    wait_duration_ms: u64,
    clock: &dyn Clock,
    sleeper: &dyn Sleeper,
) -> Result<(), String> {
    let deadline_ms = clock.now_ms().saturating_add(wait_duration_ms);
    loop {
        match process
            .child
            .lock()
            .expect("runtime client child lock should not be poisoned")
            .try_wait()
            .map_err(|error| format!("failed to poll process exit: {error}"))?
        {
            Some(_) => return Ok(()),
            None if clock.now_ms() >= deadline_ms => {
                return Err("process did not exit before stop timeout".to_string());
            }
            None => sleeper.sleep(DEFAULT_PROCESS_EXIT_POLL_INTERVAL),
        }
    }
}

/// Sends the configured termination signal to one runtime client child process.
pub(super) fn signal_runtime_client_process(
    process: &mut RunningRuntimeClientProcess,
    signal: RuntimeClientProcessStopSignal,
) -> Result<(), String> {
    let pid =
        i32::try_from(process.pid()).map_err(|_| "process pid exceeded i32 range".to_string())?;
    let unix_signal = match signal {
        RuntimeClientProcessStopSignal::Sigterm => Signal::SIGTERM,
        RuntimeClientProcessStopSignal::Sigkill => Signal::SIGKILL,
    };

    match send_signal(Pid::from_raw(pid), unix_signal) {
        Ok(()) => Ok(()),
        Err(Errno::ESRCH) => Ok(()),
        Err(error) => Err(format!(
            "failed to signal process pid={pid} signal={unix_signal}: {error}"
        )),
    }
}

/// Formats an exited child status for readiness or shutdown error reporting.
pub(super) fn describe_process_exit(status: std::process::ExitStatus) -> String {
    match status.code() {
        Some(0) => "process exited".to_string(),
        Some(code) => format!("process exited with code {code}"),
        None => "process exited with signal".to_string(),
    }
}

File: packages/sandboxd/process/lifecycle.go
package process

import (
	"fmt"
	"os/exec"
	"strings"
	"sync"
	"syscall"
	"time"

	"github.com/mistle/sandboxd/runtime"
	"github.com/mistle/sandboxd/timeutil"
)

const (
	DefaultProcessExitPollInterval      = 25 * time.Millisecond
	DefaultProcessReadinessPollInterval = 100 * time.Millisecond
	DefaultProcessMonitorPollInterval   = 1000 * time.Millisecond
	CodexAppServerFailureThreshold      = 3
	OpenCodeServerFailureThreshold      = 3
)

type RunningRuntimeClientProcess struct {
	Spec          RuntimeClientProcessSpec
	child         *exec.Cmd
	outputCapture *OutputCapture
	waitResult    chan processWaitResult
	mutex         sync.Mutex
}

type processWaitResult struct {
	err error
}

func (process *RunningRuntimeClientProcess) PID() uint32 {
	process.mutex.Lock()
	defer process.mutex.Unlock()
	if process.child.Process == nil {
		return 0
	}
	return uint32(process.child.Process.Pid)
}

func (process *RunningRuntimeClientProcess) OutputCapture() *OutputCapture {
	return process.outputCapture
}

func StartRuntimeClientProcess(processSpec RuntimeClientProcessSpec) (*RunningRuntimeClientProcess, error) {
	child, outputCapture, err := SpawnRuntimeClientChild(processSpec)
	if err != nil {
		return nil, err
	}
	return &RunningRuntimeClientProcess{
		Spec:          processSpec,
		child:         child,
		outputCapture: outputCapture,
		waitResult:    startProcessWait(child),
	}, nil
}

func SpawnRuntimeClientChild(processSpec RuntimeClientProcessSpec) (*exec.Cmd, *OutputCapture, error) {
	if len(processSpec.Command.Args) == 0 {
		return nil, nil, fmt.Errorf("process command args must not be empty")
	}
	child := exec.Command(processSpec.Command.Args[0], processSpec.Command.Args[1:]...)
	if processSpec.Command.CWD != nil {
		child.Dir = *processSpec.Command.CWD
	}
	if len(processSpec.Command.Env) > 0 {
		child.Env = append(child.Environ(), mapEnv(processSpec.Command.Env)...)
	}

	stdout, err := child.StdoutPipe()
	if err != nil {
		return nil, nil, fmt.Errorf("failed to attach process stdout: %w", err)
	}
	stderr, err := child.StderrPipe()
	if err != nil {
		return nil, nil, fmt.Errorf("failed to attach process stderr: %w", err)
	}
	child.Stdin = nil

	if err := child.Start(); err != nil {
		return nil, nil, fmt.Errorf("failed to start process command: %w", err)
	}

	outputCapture := NewOutputCapture()
	outputCapture.RegisterCaptureReader(stdout, OutputStreamStdout)
	outputCapture.RegisterCaptureReader(stderr, OutputStreamStderr)
	return child, outputCapture, nil
}

func StopStartedProcesses(processes []*RunningRuntimeClientProcess, clock timeutil.Clock, sleeper timeutil.Sleeper) error {
	stopErrors := make([]string, 0)
	for index := len(processes) - 1; index >= 0; index-- {
		process := processes[index]
		if err := StopRuntimeClientProcess(process, clock, sleeper); err != nil {
			stopErrors = append(stopErrors, fmt.Sprintf("processKey=%s: %s", process.Spec.ProcessKey, err.Error()))
		}
	}
	if len(stopErrors) > 0 {
		return fmt.Errorf("%s", strings.Join(stopErrors, "; "))
	}
	return nil
}

func StopRuntimeClientProcess(process *RunningRuntimeClientProcess, clock timeutil.Clock, sleeper timeutil.Sleeper) error {
	exited, err := ProcessHasExited(process)
	if err != nil {
		return err
	}
	if exited {
		process.outputCapture.FinishCaptureReaders()
		return nil
	}

	deadlineMS := clock.NowMS() + process.Spec.Stop.TimeoutMS
	if err := SignalRuntimeClientProcess(process, process.Spec.Stop.Signal); err != nil {
		return err
	}

	if process.Spec.Stop.Signal == runtime.RuntimeClientProcessStopSignalSIGTERM {
		gracePeriodMS := uint64(0)
		if process.Spec.Stop.GracePeriodMS != nil {
			gracePeriodMS = *process.Spec.Stop.GracePeriodMS
		}
		if gracePeriodMS > 0 {
			if err := WaitForRuntimeClientProcessExit(process, gracePeriodMS, clock, sleeper); err == nil {
				process.outputCapture.FinishCaptureReaders()
				return nil
			}
			if err := SignalRuntimeClientProcess(process, runtime.RuntimeClientProcessStopSignalSIGKILL); err != nil {
				return err
			}
		}
	}

	remainingMS := deadlineMS - clock.NowMS()
	if clock.NowMS() > deadlineMS {
		remainingMS = 0
	}
	err = WaitForRuntimeClientProcessExit(process, remainingMS, clock, sleeper)
	if err == nil {
		process.outputCapture.FinishCaptureReaders()
	}
	return err
}

func ProcessHasExited(process *RunningRuntimeClientProcess) (bool, error) {
	select {
	case <-process.waitResult:
		process.waitResult = closedProcessWaitResult()
		return true, nil
	default:
		return false, nil
	}
}

func WaitForRuntimeClientProcessExit(
	process *RunningRuntimeClientProcess,
	waitDurationMS uint64,
	clock timeutil.Clock,
	sleeper timeutil.Sleeper,
) error {
	deadlineMS := clock.NowMS() + waitDurationMS
	for {
		exited, err := reapExitedProcess(process)
		if err != nil {
			return err
		}
		if exited {
			return nil
		}
		if clock.NowMS() >= deadlineMS {
			return fmt.Errorf("process did not exit before stop timeout")
		}
		sleeper.Sleep(DefaultProcessExitPollInterval)
	}
}

func WaitForRuntimeClientProcessReadiness(
	process *RunningRuntimeClientProcess,
	clock timeutil.Clock,
	sleeper timeutil.Sleeper,
) error {
	if process.Spec.Readiness.Type == runtime.RuntimeClientProcessReadinessNone {
		return nil
	}
	timeoutMS := ReadinessTimeoutMS(process.Spec)
	deadlineMS := clock.NowMS() + timeoutMS

	for {
		exited, err := ProcessHasExited(process)
		if err != nil {
			return err
		}
		if exited {
			return fmt.Errorf("%s", processExitDescription(process))
		}

		readinessErr := CheckRuntimeClientProcessReadinessFromSpec(process.Spec)
		if readinessErr == nil {
			return nil
		}
		if clock.NowMS() >= deadlineMS {
			return fmt.Errorf(
				"timed out after %dms waiting for readiness: %w",
				timeoutMS,
				readinessErr,
			)
		}

		sleeper.Sleep(DefaultProcessReadinessPollInterval)
	}
}

func SignalRuntimeClientProcess(process *RunningRuntimeClientProcess, signal runtime.RuntimeClientProcessStopSignal) error {
	process.mutex.Lock()
	defer process.mutex.Unlock()
	if process.child.Process == nil {
		return nil
	}
	unixSignal, err := unixStopSignal(signal)
	if err != nil {
		return err
	}
	pid := process.child.Process.Pid
	if err := process.child.Process.Signal(unixSignal); err != nil {
		if err == syscall.ESRCH {
			return nil
		}
		return fmt.Errorf("failed to signal process pid=%d signal=%s: %w", pid, unixSignal.String(), err)
	}
	return nil
}

func DescribeProcessExit(processState osProcessState) string {
	return describeProcessExitStatus(processState.ExitCode())
}

func processExitEventFields(process *RunningRuntimeClientProcess) (string, map[string]any) {
	process.mutex.Lock()
	defer process.mutex.Unlock()
	processState := process.child.ProcessState
	if processState == nil {
		return "process_exited", map[string]any{
			"exitKind": "process_exited",
			"exitCode": 0,
		}
	}
	waitStatus, ok := processState.Sys().(syscall.WaitStatus)
	if ok && waitStatus.Signaled() {
		return "process_signaled", map[string]any{
			"exitKind": "process_signaled",
			"signal":   int(waitStatus.Signal()),
		}
	}
	return "process_exited", map[string]any{
		"exitKind": "process_exited",
		"exitCode": processState.ExitCode(),
	}
}

func describeProcessExitStatus(exitCode int) string {
	switch exitCode {
	case 0:
		return "process exited"
	case -1:
		return "process exited with signal"
	default:
		return fmt.Sprintf("process exited with code %d", exitCode)
	}
}

type osProcessState interface {
	ExitCode() int
	String() string
}

func reapExitedProcess(process *RunningRuntimeClientProcess) (bool, error) {
	select {
	case result := <-process.waitResult:
		process.waitResult = closedProcessWaitResult()
		return waitResultExited(result)
	default:
		return false, nil
	}
}

func processExitDescription(process *RunningRuntimeClientProcess) string {
	process.mutex.Lock()
	defer process.mutex.Unlock()
	if process.child.ProcessState == nil {
		return "process exited"
	}
	return DescribeProcessExit(process.child.ProcessState)
}

func startProcessWait(child *exec.Cmd) chan processWaitResult {
	result := make(chan processWaitResult, 1)
	go func() {
		result <- processWaitResult{err: child.Wait()}
		close(result)
	}()
	return result
}

func closedProcessWaitResult() chan processWaitResult {
	channel := make(chan processWaitResult)
	close(channel)
	return channel
}

func waitResultExited(result processWaitResult) (bool, error) {
	if result.err == nil {
		return true, nil
	}
	if _, ok := result.err.(*exec.ExitError); ok {
		return true, nil
	}
	return false, fmt.Errorf("failed to poll process exit: %w", result.err)
}

func unixStopSignal(signal runtime.RuntimeClientProcessStopSignal) (syscall.Signal, error) {
	switch signal {
	case runtime.RuntimeClientProcessStopSignalSIGTERM:
		return syscall.SIGTERM, nil
	case runtime.RuntimeClientProcessStopSignalSIGKILL:
		return syscall.SIGKILL, nil
	default:
		return 0, fmt.Errorf("unsupported process stop signal: %s", signal)
	}
}

func mapEnv(env map[string]string) []string {
	result := make([]string, 0, len(env))
	for name, value := range env {
		result = append(result, name+"="+value)
	}
	return result
}
